examples/api/README.md

```diff
@@ -1,67 +1,125 @@
-/// ```
+/// ** See code in examples/api/lib/animation/curves/curve2_d.0.dart **
 /// {@end-tool}
 ```
 
-Then install the `extract_sample` command with:
+The "See code" link needs to be formatted exactly as above, with no wrapping or
+newlines, one space after the "`**`" at the beginning, and one space before the
+"`**`" at the end, and the words "See code" at the beginning of the line.  This
+is what the snippets tool and the IDE use when finding the example source code
+that you are creating.
 
-```
-% pub global activate snippets
-```
+Use `{@tool dartpad}` for Dartpad examples, and use `{@tool sample}` for
+examples that shouldn't be run/shown in Dartpad.
 
-And run the `extract_sample` command from the Flutter repo dir:
+Once that comment block is inserted in the source code, create a new file at the
+appropriate path under [`examples/api`](.). You should also add tests for your
+sample code under [`examples/api/test`](./test).
 
+The entire example should be in a single file, so that Dartpad can load it.
+
+Only packages that can be loaded by Dartpad may be imported. If you use one that
+hasn't been used in an example before, you may have to add it to the
+[pubspec.yaml](pubspec.yaml) in the api directory.
+
+## Snippets
+
+There is another type of example that can also be authored, using `{@tool snippet}`. Snippet examples are just written inline in the source, like so:
+
+```dart
+/// {@tool dartpad}
+/// Write a description of the example here. This description will appear in the
+/// API web documentation to introduce the example.
+///
+/// ```dart
+/// // Sample code goes here, e.g.:
+/// const Widget emptyBox = SizedBox();
+/// ```
+/// {@end-tool}
 ```
-$ pub global run extract_sample packages/flutter/lib/src/animation/curves.dart
-```
 
-This will create a new file in the `examples/api` directory, in this case it
-would create `examples/api/lib/animation/curves/curve2_d.1.dart`
+The source for these snippets isn't stored under the [`examples/api`](.)
+directory, or available in Dartpad in the API docs, since they're not intended
+to be runnable, they just show some incomplete snippet of example code. It must
+compile (in the context of the sample analyzer), but doesn't need to do
+anything. See [the snippets
+documentation](https://pub.dev/packages/snippets#snippet-tool) for more
+information about the context that the analyzer uses.
+
+## Writing Tests
+
+Examples should have tests. There is already a "smoke test" that runs all the
+api examples, just to make sure that they start up without crashing. Beyond
+that, it is possible to write tests of functionality in the examples, capture
+golden tests, and generally just do what we normally do for unit tests. The one
+thing that makes it more challenging is that the examples can't really be
+written for testability in any obvious way, since that would probably complicate
+the example and make it harder to explain. You can still do a lot though, with
```

Comment:
Not sure what "with composition and overrides" refers to here.  Can we test layout and (input) behavior?